File: src/main/java/me/redth/statsoverlay/stat/bedwars/BedwarsPrestige.java
package me.redth.statsoverlay.stat.bedwars;

public enum BedwarsPrestige {
    STONE("§7[x✫]", false),
    IRON("§f[x✫]", false),
    GOLD("§6[x✫]", false),
    DIAMOND("§b[x✫]", false),
    EMERALD("§2[x✫]", false),
    SAPPHIRE("§3[x✫]", false),
    RUBY("§4[x✫]", false),
    CRYSTAL("§d[x✫]", false),
    OPAL("§9[x✫]", false),
    AMETHYST("§5[x✫]", false),
    RAINBOW("§c[§6w§ex§ay§bz§d✫§5]", true),
    IRON_PRIME("§7[§fx§7✪§7]", false),
    GOLD_PRIME("§7[§ex§6✪§7]", false),
    DIAMOND_PRIME("§7[§bx§3✪§7]", false),
    EMERALD_PRIME("§7[§ax§2✪§7]", false),
    SAPPHIRE_PRIME("§7[§3x§9✪§7]", false),
    RUBY_PRIME("§7[§cx§4✪§7]", false),
    CRYSTAL_PRIME("§7[§dx§5✪§7]", false),
    OPAL_PRIME("§7[§9x§1✪§7]", false),
    AMETHYST_PRIME("§7[§5x§8✪§7]", false),
    MIRROR("§8[§7w§fxy§7z§8✪]", true),
    LIGHT("§f[w§exy§6z§l⚝§6]", true),
    DAWN("§6[w§fxy§bz§3§l⚝§3]", true),
    DUSK("§5[w§dxy§6z§e§l⚝§e]", true),
    AIR("§b[w§fxy§7z§l⚝§8]", true),
    WIND("§f[w§axy§2z§l⚝§2]", true),
    NEBULA("§4[w§cxy§dz§l⚝§d]", true),
    THUNDER("§e[w§fxy§8z§l⚝§8]", true),
    EARTH("§a[w§2xy§6z§l⚝§e]", true),
    WATER("§b[w§3xy§9z§l⚝§1]", true),
    FIRE("§e[w§6xy§cz§l⚝§4]", true);

    private final String format;
    private final boolean splitNumbers;

    BedwarsPrestige(String format, boolean splitNumbers) {
        this.format = format;
        this.splitNumbers = splitNumbers;
    }

    public static String format(int stars) {
        BedwarsPrestige prestige = getPrestige(stars);
        String starsString = String.valueOf(stars);
        if (prestige.splitNumbers)
            return prestige.format.replace('w', starsString.charAt(0)).replace('x', starsString.charAt(1)).replace('y', starsString.charAt(2)).replace('z', starsString.charAt(3));
        else
            return prestige.format.replace("x", starsString);

    }

    public static BedwarsPrestige getPrestige(int stars) {
        if (stars < 0 || stars / 100 >= values().length) return STONE;
        return values()[stars / 100];
    }
}

File: src/main/kotlin/me/redth/statsoverlay/bedwars/BedwarsPrestige.kt
package me.redth.statsoverlay.bedwars

import me.redth.statsoverlay.data.ColoredText

enum class BedwarsPrestige(
    private val format: String,
) {
    STONE("§7[_✫]"),
    IRON("§f[_✫]"),
    GOLD("§6[_✫]"),
    DIAMOND("§b[_✫]"),
    EMERALD("§2[_✫]"),
    SAPPHIRE("§3[_✫]"),
    RUBY("§4[_✫]"),
    CRYSTAL("§d[_✫]"),
    OPAL("§9[_✫]"),
    AMETHYST("§5[_✫]"),
    RAINBOW("§c[§6_§e_§a_§b_§d✫§5]"),
    IRON_PRIME("§7[§f_§7✪§7]"),
    GOLD_PRIME("§7[§e_§6✪§7]"),
    DIAMOND_PRIME("§7[§b_§3✪§7]"),
    EMERALD_PRIME("§7[§a_§2✪§7]"),
    SAPPHIRE_PRIME("§7[§3_§9✪§7]"),
    RUBY_PRIME("§7[§c_§4✪§7]"),
    CRYSTAL_PRIME("§7[§d_§5✪§7]"),
    OPAL_PRIME("§7[§9_§1✪§7]"),
    AMETHYST_PRIME("§7[§5_§8✪§7]"),
    MIRROR("§8[§7_§f__§7_§8✪]"),
    LIGHT("§f[_§e__§6_§l⚝§6]"),
    DAWN("§6[_§f__§b_§3§l⚝§3]"),
    DUSK("§5[_§d__§6_§e§l⚝§e]"),
    AIR("§b[_§f__§7_§l⚝§8]"),
    WIND("§f[_§a__§2_§l⚝§2]"),
    NEBULA("§4[_§c__§d_§l⚝§d]"),
    THUNDER("§e[_§f__§8_§l⚝§8]"),
    EARTH("§a[_§2__§6_§l⚝§e]"),
    WATER("§b[_§3__§9_§l⚝§1]"),
    FIRE("§e[_§6__§c_§l⚝§4]");

    private val shouldSplitDigits: Boolean = format.count { it == '_' } > 1

    companion object {
        fun getPrefix(stars: Int) = ColoredText(format(stars) + " §r")

        fun format(stars: Int): String {
            val prestige = getPrestige(stars)
            val starsString = stars.toString()

            if (!prestige.shouldSplitDigits) return prestige.format.replaceFirst("_", starsString)

            return prestige.format
                .replaceFirst('_', starsString[0])
                .replaceFirst('_', starsString[1])
                .replaceFirst('_', starsString[2])
                .replaceFirst('_', starsString[3])
        }

        private fun getPrestige(stars: Int) = entries.getOrNull(stars / 100) ?: STONE
    }
}
